pkg/landscaper/crdmanager/crdmanager.go

```diff
@@ -0,0 +1,186 @@
+		return nil
+	}
+
+	crdList, err := crdmgr.crdsFromDir()
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	defer ctx.Done()
+
+	crdmgr.log.V(1).Info("Registering Landscaper CRDs in cluster")
+
+	for _, crd := range crdList {
+		existingCrd, err := crdmgr.crdClientset.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, crd.Name, metav1.GetOptions{TypeMeta: crd.TypeMeta})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				_, err := crdmgr.createCrd(ctx, &crd)
+				if err != nil {
+					return err
+				}
+
+				continue
+			}
+
+			return err
+		}
+
+		_, err = crdmgr.updateCrd(ctx, existingCrd, &crd)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = wait.Poll(1*time.Second, 30*time.Second, func() (bool, error) {
+		aggregatedStatus := true
+
+		for _, crd := range crdList {
+			if aggregatedStatus == false {
+				return aggregatedStatus, nil
+			}
+
+			crdResult, err := crdmgr.crdClientset.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, crd.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+
+			for _, crdCondition := range crdResult.Status.Conditions {
+				switch crdCondition.Type {
+				case v1.Established:
+					if crdCondition.Status != v1.ConditionTrue {
+						aggregatedStatus = false
+					}
+				case v1.NamesAccepted:
+					if crdCondition.Status == v1.ConditionFalse {
+						aggregatedStatus = false
+					}
+				}
+			}
+		}
+		return aggregatedStatus, nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (crdmgr *CRDManager) createCrd(ctx context.Context, crd *v1.CustomResourceDefinition) (*v1.CustomResourceDefinition, error) {
+	return crdmgr.crdClientset.ApiextensionsV1().CustomResourceDefinitions().Create(ctx, crd, metav1.CreateOptions{TypeMeta: crd.TypeMeta})
+}
+
+func (crdmgr *CRDManager) updateCrd(ctx context.Context, currentCrd, updatedCrd *v1.CustomResourceDefinition) (*v1.CustomResourceDefinition, error) {
+	if !crdmgr.cfg.ForceUpdate {
+		return nil, nil
+	}
+
+	updatedCrd.ResourceVersion = currentCrd.ResourceVersion
+	updatedCrd.UID = currentCrd.UID
+	return crdmgr.crdClientset.ApiextensionsV1().CustomResourceDefinitions().Update(ctx, updatedCrd, metav1.UpdateOptions{TypeMeta: updatedCrd.TypeMeta})
+}
+
+func (crdmgr *CRDManager) crdsFromDir() ([]v1.CustomResourceDefinition, error) {
+	crdList := make([]v1.CustomResourceDefinition, 0)
+
+	files, err := crdmgr.crdRawDataFS.ReadDir(embedFSCrdRootDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		data, err := crdmgr.crdRawDataFS.ReadFile(path.Join(embedFSCrdRootDir, file.Name()))
+		if err != nil {
+			crdmgr.log.V(3).Error(err, "failed to read CRD file")
+			return nil, err
+		}
+
+		decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(data), readerBufferSize)
+		crd := &v1.CustomResourceDefinition{}
+		err = decoder.Decode(crd)
+		if err != nil {
+			crdmgr.log.V(3).Error(err, "failed to decode CRD from file")
+			return nil, err
```

Comment:
the err should be logged by the caller and therefore it would be logged 2 times.
```suggestion
			return nil, fmt.Errorf("failed to decode CRD from file: %w", err)
```

Reply:
Fixed.